Chore: Fully deprecate `objc2::rc::Id`

PaulDance opened the pull request:
As part of #617.

With this, 0.6 should be in a good position to be delivered, or are there still a lot of things to do? What I see in https://github.com/madsmtm/objc2/milestone/6 seems that it could be optional while the important changes that are already completed would be nice if published and back-porting things I would need in a fork would be cumbersome.

madsmtm reviewed (changes requested):
> As far as I can tell, this should be the last remaining part required in order to close #617.

I think I'd also like to figure out `#[method_id(...)]`, I've added mention of https://github.com/madsmtm/objc2/issues/457 to the issue description.

> With this, 0.6 should be in a good position to be delivered, or are there still a lot of things to do? What I see in https://github.com/madsmtm/objc2/milestone/6 seems that it could be optional while the important changes that are already completed would be nice if published and back-porting things I would need in a fork would be cumbersome.

Ugh, I'm so bad at deciding when to release, in my defence things are complicated by us having a fairly large amount of dependents, and thus infrequent breaking versions are desirable for the ecosystem in general. But you are right, we are _close_, the target date I've set for the milestone (end of the year) _should_ be accurate now, the remaining stuff that I really want to get in is:
- Proper enum prefix stripping, having to write `NSWindowSharingType::NSWindowSharingNone` is really ugly when `NSWindowSharingType::None` could do.
- The cargo features stuff (removing `"all"` in favour of enabling things by default + std features shouldn't enable sub-features), since it's a breaking change.
- The autogenerated thunks for functions, since it's also a breaking change, and it's needed for nicer Metal integration (allows `objc2-metal` to really shine over `metal-rs`).

(So if you want to explore some of those, then feel free to ;) )

PaulDance commented on crates/objc2/src/rc/retained.rs:
```diff
@@ -146,11 +146,8 @@ pub struct Retained<T: ?Sized> {
-/// [#617]: https://github.com/madsmtm/objc2/issues/617
+/// This is fully deprecated since `v0.6.0`, use [`Retained`] instead.
+#[deprecated(since = "0.6.0", note = "Renamed to `Retained`.")]
 pub type Id<T> = Retained<T>;
```
Didn't notice this one! Done :slightly_smiling_face:

madsmtm reviewed (approved):
Thanks!

PaulDance commented:
> I think I'd also like to figure out `#[method_id(...)]`, I've added mention of #457 to the issue description.

OK, changed the description accordingly.

> I'm so bad at deciding when to release

Most definitely understandable. Another way to see things though is that breaking a lot at the same time makes transitioning harder, while more frequent but more focused breakage is somewhat more manageable. Tough choices overall.

> the target date I've set for the milestone (end of the year) should be accurate now

That is more than acceptable, especially considering vacations and stuff. It should also align nicely with a development I will take on again.

> So if you want to explore some of those, then feel free to ;)

I just might, but maybe only the first two :wink:

> Thanks!

You're welcome!

madsmtm commented:
Update on the timeline I gave: I of course didn't make it by the end of the year, but I _have_ progressed on the things I said I wanted to get in, the main things missing is the cargo features stuff (I have a local WIP branch for this, but it's tricky). So we are getting _closer_...

PaulDance commented:
Yes, I have seen the recent progress and it's looking good indeed! Regarding the timeline, the projects were big and numerous, so it is not exactly surprising the originally-estimated date was passed. However, considering it should only be by a few days in total and that my use of it is only optional during the development that is only starting now really, I think it should be fine :slightly_smiling_face:
